### PR TITLE
ICACHE_RAM_ATTR attribute fix for esp8266 and esp32 boards

### PR DESCRIPTION
**Pull request by Di-Strix**
Added ICACHE_RAM_ATTR attribute to isr functions for esp8266 and esp32 boards.
Added yield() to waitForData function to avoid WDT reset on esp boards.
Fixes issue #9

### CODE REVIEW

**Comment by strange-v** [comment]
Thank you for contributing, could you fix a few things?

1. Please use a similar approach to this https://github.com/strange-v/FT6X36/blob/7507eea1128b805508775894d6ad125f2d5705fd/FT6X36.h#L7
2. I know that for ESP32 ICACHE_RAM_ATTR is defined as IRAM_ATTR in esp8266-compat.h, but I'm not sure that this backward compatibility fix is permanent. Please, use IRAM_ATTR for ESP32.

**Comment by Di-Strix** [comment]
@strange-v Done

File: MHZ19PWM.h [hunk 1]
```diff
@@ -1,6 +1,14 @@
 #pragma once
 #include <Arduino.h>
 
+#if defined(ARDUINO_ARCH_ESP8266)
+#define ISR_FN_ATTR ICACHE_RAM_ATTR
```
Comment (strange-v) [inline MHZ19PWM.h]: Please, use ISR_ATTR in all places instead of ISR_FN_ATTR.

File: MHZ19PWM.h [hunk 2]
```diff
@@ -11,6 +19,7 @@ class MHZ19PWM
 {
 public:
 	static MHZ19PWM * instance;
+
```
Comment (strange-v) [inline MHZ19PWM.h]: Remove newline

File: MHZ19PWM.h [hunk 3]
```diff
@@ -34,5 +43,7 @@ class MHZ19PWM
 	void start();
 	void stop();
 	void waitForData();
+
```
Comment (strange-v) [inline MHZ19PWM.h]: Remove newline

File: MHZ19PWM.h [hunk 4]
```diff
@@ -34,5 +43,7 @@ class MHZ19PWM
 	void start();
 	void stop();
 	void waitForData();
+
 	void isrInternal();
+
```
Comment (strange-v) [inline MHZ19PWM.h]: Remove newline

**Review by strange-v — COMMENTED** [review]
A few more minor changes (commented).
Is it okay to have IRAM_ATTR only in definition (.cpp) or should it be in the declaration too (.h)? Honestly, I don't remember.

**Comment by Di-Strix** [comment]
> A few more minor changes (commented).
> Is it okay to have IRAM_ATTR only in definition (.cpp) or should it be in the declaration too (.h)? Honestly, I don't remember.

Because of you're including header file in cpp file then all contents of both files are "merging" together. So you need to define or in the header file(.h) or in the cpp file. In most cases all defines are in the header file, but nothing is stopping you from moving them to the cpp file.

Move IRAM_ATTR define to the MHZ19PWM.cpp?

**Comment by strange-v** [comment]
No, defines should be in the header, it's logical.
I'm talking about attributing the method, for example, MHZ19PWM::isr() attributed (ISR_FN_ATTR) in the cpp file, but not in the header.
If it is possible to attribute only in one place I'd prefer the header file.

**Comment by Di-Strix** [comment]
I just tested and seems it is worked. So I'm moving ISR_ATTR to the header

**Comment by strange-v** [comment]
Thank you!

**Pull request merged**